qemu: Disable features/devices per default

zvonkok opened the pull request:
We have a very long list of disable calls for QEMU that we can replace by using the configure script to disable all default features and default devices.
We enable selectively the features that we need.

zvonkok commented:
Went down from `[1/2085]` to `[1/1585]` compiled units.

zvonkok commented:
Well looks like that we still need the disable calls since some of those features are not default
```
qemu_options+=(minimal:--without-default-features)
qemu_options+=(minimal:--without-default-devices)
```
CI is green need to test the GPU to see if this works.

zvonkok commented:
Build log: 
```
qemu 9.1.2

  Build environment
    Build directory                              : /zvonkok/github.com/zvonkok/kata-containers/tools/packaging/kata-deploy/local-build/build/qemu/builddir/qemu/build
    Source path                                  : /zvonkok/github.com/zvonkok/kata-containers/tools/packaging/kata-deploy/local-build/build/qemu/builddir/qemu
    Download dependencies                        : YES

  Directories
    Build directory                              : /zvonkok/github.com/zvonkok/kata-containers/tools/packaging/kata-deploy/local-build/build/qemu/builddir/qemu/build
    Source path                                  : /zvonkok/github.com/zvonkok/kata-containers/tools/packaging/kata-deploy/local-build/build/qemu/builddir/qemu
    Download dependencies                        : YES
    Install prefix                               : /opt/kata
    BIOS directory                               : share/kata-qemu/qemu
    firmware path                                : share/qemu-firmware
    binary directory                             : /opt/kata/bin
    library directory                            : /opt/kata/lib/kata-qemu
    module directory                             : lib/kata-qemu/qemu
    libexec directory                            : /opt/kata/libexec/kata-qemu
    include directory                            : /opt/kata/include
    config directory                             : /opt/kata/etc
    local state directory                        : /opt/kata/var
    Manual directory                             : /opt/kata/share/man
    Doc directory                                : /opt/kata/share/doc

  Host binaries
    python                                       : /zvonkok/github.com/zvonkok/kata-containers/tools/packaging/kata-deploy/local-build/build/qemu/builddir/qemu/build/pyvenv/bin/python3 (version: 3.10)
    sphinx-build                                 : NO
    gdb                                          : 
    iasl                                         : NO
    genisoimage                                  : 

  Configurable features
    Documentation                                : NO
    system-mode emulation                        : YES
    user-mode emulation                          : NO
    block layer                                  : YES
    Install blobs                                : YES
    module support                               : NO
    fuzzing support                              : NO
    Audio drivers                                : 
    Trace backends                               : log
    D-Bus display                                : NO
    QOM debugging                                : NO
    Relocatable install                          : NO
    vhost-kernel support                         : NO
    vhost-net support                            : NO
    vhost-user support                           : NO
    vhost-user-crypto support                    : NO
    vhost-user-blk server support                : NO
    vhost-vdpa support                           : NO
    build guest agent                            : NO

  Compilation
    host CPU                                     : x86_64
    host endianness                              : little
    C compiler                                   : cc -m64
    Host C compiler                              : cc -m64
    C++ compiler                                 : NO
    Objective-C compiler                         : NO
    CFLAGS                                       : -O2 -fno-semantic-interposition -falign-functions=32 -D_FORTIFY_SOURCE=2 -g -O2
    LDFLAGS                                      : -O2 -fno-semantic-interposition -falign-functions=32 -D_FORTIFY_SOURCE=2 -z noexecstack -z relro -z now
    QEMU_CFLAGS                                  : -mcx16 -msse2 -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -fno-strict-aliasing -fno-common -fwrapv -fzero-call-used-regs=used-gpr
    QEMU_LDFLAGS                                 : -static-pie -Wl,-z,relro -Wl,-z,now
    link-time optimization (LTO)                 : NO
    PIE                                          : YES
    static build                                 : YES
    malloc trim support                          : YES
    membarrier                                   : NO
    debug graph lock                             : NO
    debug stack usage                            : NO
    mutex debugging                              : NO
    memory allocator                             : system
    avx2 optimization                            : YES
    avx512bw optimization                        : YES
    gcov                                         : NO
    thread sanitizer                             : NO
    CFI support                                  : NO
    strip binaries                               : NO
    sparse                                       : NO
    mingw32 support                              : NO

  Cross compilers
    x86_64                                       : cc

  Targets and accelerators
    KVM support                                  : YES
    HVF support                                  : NO
    WHPX support                                 : NO
    NVMM support                                 : NO
    Xen support                                  : NO
    Xen emulation                                : NO
    TCG support                                  : NO
    target list                                  : x86_64-softmmu
    default devices                              : NO
    out of process emulation                     : NO
    vfio-user server                             : NO

  Block layer support
    coroutine backend                            : ucontext
    coroutine pool                               : YES
    Block whitelist (rw)                         : 
    Block whitelist (ro)                         : 
    Use block whitelist in tools                 : NO
    VirtFS (9P) support                          : YES
    VirtFS (9P) Proxy Helper support (deprecated): NO
    replication support                          : NO
    bochs support                                : NO
    cloop support                                : NO
    dmg support                                  : NO
    qcow v1 support                              : NO
    vdi support                                  : NO
    vhdx support                                 : NO
    vmdk support                                 : NO
    vpc support                                  : NO
    vvfat support                                : NO
    qed support                                  : NO
    parallels support                            : NO
    FUSE exports                                 : NO
    VDUSE block exports                          : NO

  Crypto
    TLS priority                                 : NORMAL
    GNUTLS support                               : NO
    libgcrypt                                    : NO
    nettle                                       : NO
    SM4 ALG support                              : NO
    AF_ALG support                               : NO
    rng-none                                     : NO
    Linux keyring                                : NO
    Linux keyutils                               : NO

  User interface
    SDL support                                  : NO
    SDL image support                            : NO
    GTK support                                  : NO
    pixman                                       : NO
    VTE support                                  : NO
    PNG support                                  : NO
    VNC support                                  : NO
    spice protocol support                       : NO
    curses support                               : NO
    brlapi support                               : NO

  Graphics backends
    VirGL support                                : NO
    Rutabaga support                             : NO

  Audio backends
    OSS support                                  : NO
    sndio support                                : NO
    ALSA support                                 : NO
    PulseAudio support                           : NO
    PipeWire support                             : NO
    JACK support                                 : NO

  Network backends
    AF_XDP support                               : NO
    slirp support                                : NO
    vde support                                  : NO
    netmap support                               : NO
    l2tpv3 support                               : NO

  Dependencies
    libtasn1                                     : NO
    PAM                                          : NO
    iconv support                                : NO
    blkio support                                : NO
    curl support                                 : NO
    Multipath support                            : NO
    Linux AIO support                            : YES
    Linux io_uring support                       : YES 2.0
    ATTR/XATTR support                           : YES
    RDMA support                                 : NO
    fdt support                                  : internal
    libcap-ng support                            : YES
    bpf support                                  : NO
    rbd support                                  : NO
    smartcard support                            : NO
    U2F support                                  : NO
    libusb                                       : NO
    usb net redir                                : NO
    OpenGL support (epoxy)                       : NO
    GBM                                          : NO
    libiscsi support                             : NO
    libnfs support                               : NO
    seccomp support                              : YES 2.5.3
    GlusterFS support                            : NO
    hv-balloon support                           : NO
    TPM support                                  : NO
    libssh support                               : NO
    lzo support                                  : NO
    snappy support                               : NO
    bzip2 support                                : NO
    lzfse support                                : NO
    zstd support                                 : NO
    Query Processing Library support             : NO
    UADK Library support                         : NO
    NUMA host support                            : NO
    capstone                                     : NO
    libpmem support                              : NO
    libdaxctl support                            : NO
    libudev                                      : NO
    FUSE lseek                                   : NO
    selinux                                      : NO
    libdw                                        : NO

  Subprojects
    dtc                                          : YES
    keycodemapdb                                 : YES

  User defined options
    Native files                                 : config-meson.cross
    auto_features                                : disabled
    datadir                                      : /opt/kata/share/kata-qemu
    libdir                                       : /opt/kata/lib/kata-qemu
    libexecdir                                   : /opt/kata/libexec/kata-qemu
    prefer_static                                : true
    prefix                                       : /opt/kata
    attr                                         : enabled
    auth_pam                                     : disabled
    avx2                                         : enabled
    avx512bw                                     : enabled
    bochs                                        : disabled
    brlapi                                       : disabled
    bzip2                                        : disabled
    cap_ng                                       : enabled
    capstone                                     : disabled
    cloop                                        : disabled
    colo_proxy                                   : disabled
    curl                                         : disabled
    curses                                       : disabled
    debug_graph_lock                             : false
    debug_remap                                  : false
    debug_tcg                                    : false
    default_devices                              : false
    dmg                                          : disabled
    docs                                         : disabled
    gio                                          : disabled
    glusterfs                                    : disabled
    gtk                                          : disabled
    guest_agent                                  : disabled
    guest_agent_msi                              : disabled
    hexagon_idef_parser                          : false
    hv_balloon                                   : disabled
    kvm                                          : enabled
    libdaxctl                                    : disabled
    libdw                                        : disabled
    libiscsi                                     : disabled
    libnfs                                       : disabled
    libpmem                                      : disabled
    libssh                                       : disabled
    libudev                                      : disabled
    libusb                                       : disabled
    linux_aio                                    : enabled
    linux_io_uring                               : enabled
    lzo                                          : disabled
    malloc_trim                                  : enabled
    nettle                                       : disabled
    opengl                                       : disabled
    oss                                          : disabled
    parallels                                    : disabled
    pipewire                                     : disabled
    pixman                                       : disabled
    pkgversion                                   : kata-static
    png                                          : disabled
    qcow1                                        : disabled
    qed                                          : disabled
    qom_cast_debug                               : false
    qpl                                          : disabled
    rdma                                         : disabled
    relocatable                                  : false
    replication                                  : disabled
    rutabaga_gfx                                 : disabled
    sdl                                          : disabled
    seccomp                                      : enabled
    slirp                                        : disabled
    smartcard                                    : disabled
    snappy                                       : disabled
    sparse                                       : disabled
    spice                                        : disabled
    tcg                                          : disabled
    tcg_interpreter                              : false
    tools                                        : disabled
    uadk                                         : disabled
    usb_redir                                    : disabled
    vde                                          : disabled
    vdi                                          : disabled
    vhdx                                         : disabled
    vhost_net                                    : enabled
    virglrenderer                                : disabled
    virtfs                                       : enabled
    vmdk                                         : disabled
    vnc                                          : disabled
    vnc_jpeg                                     : disabled
    vnc_sasl                                     : disabled
    vpc                                          : disabled
    vte                                          : disabled
    vvfat                                        : disabled
    xen                                          : disabled
```

gkurz reviewed (approved):
/lgtm

Thanks @zvonkok !

FWIW I could comment out nearly all the `size:--disable-*` options on top of this PR and I got the same number of compiled units `[1/1585]`. Future work could be to make `tools/packaging/scripts/configure-hypervisor.sh` leaner.

zvonkok commented:
@gkurz This was the intention to remove all the disable calls, but we may need to keep track if options are enabled for other release and then we would need to add disable calls again.

gkurz commented:
> @gkurz This was the intention to remove all the disable calls, but we may need to keep track if options are enabled for other release and then we would need to add disable calls again.

@zvonkok Not sure to understand what `other release` is referring to. If this is some other release of QEMU, we just care for what we have in `versions.yaml`, right ? Why should we keep disable stanzas if they are already handled ?

zvonkok commented:
@gkurz Not all disable options that we have in the script are handled by `--without-default-features` e,g, TCG is enabled with this option so we need an explicit call to disable it. 
My point was that some features move out of the default category and we need to add an extra disable option.

zvonkok commented:
Hold on... all the required tests are passing even there is no q35 machine type available in this QEMU build. 
```
time="2025-01-24T18:48:08.636608719Z" level=error msg="qemu-system-x86_64: unsupported machine type: \"q35\"" name=containerd-shim-v2 pid=1378391 qemuPid=1378402 sandbox=22b161534332819d9facdaa7e75efb02150d01ea14fa15564298996481ccb2e6 source=virtcontainers/hypervisor subsystem=qemu
``` 
How in the cinnamon toast is this working? @gkurz @fidencio

zvonkok commented:
@gkurz 
```
+ _array=('' 'minimal:--without-default-features' 'minimal:--without-default-devices' 'size:--disable-brlapi' 'minimal:--disable-docs' 'size:--disable-curses' 'size:--disable-gtk' 'size:--disable-opengl' 'size:--disable-sdl' 'size:--disable-spice' 'size:--disable-vte' 'size:--disable-vnc' 'size:--disable-vnc-jpeg' 'size:--disable-png' 'size:--disable-vnc-sasl' 'size:--disable-auth-pam' 'size:--disable-glusterfs' 'size:--disable-libiscsi' 'size:--disable-libnfs' 'size:--disable-libssh' 'size:--disable-bzip2' 'size:--disable-lzo' 'size:--disable-snappy' 'size:--disable-slirp' 'size:--disable-libusb' 'size:--disable-usb-redir' 'size:--disable-tcg' 'misc:--static' 'size:--disable-debug-tcg' 'size:--disable-tcg-interpreter' 'size:--disable-qom-cast-debug' 'size:--disable-libudev' 'security:--disable-curl' 'size:--disable-rdma' 'size:--disable-tools' 'functionality:--enable-virtfs' 'size:--disable-bsd-user' 'size:--disable-linux-user' 'size:--disable-sparse' 'size:--disable-vde' 'size:--disable-nettle' 'size:--disable-xen' 'size:--disable-capstone' 'size:--disable-virglrenderer' 'size:--disable-replication' 'size:--disable-smartcard' 'size:--disable-guest-agent' 'size:--disable-guest-agent-msi' 'size:--disable-vvfat' 'size:--disable-vdi' 'size:--disable-qed' 'size:--disable-qcow1' 'size:--disable-bochs' 'size:--disable-cloop' 'size:--disable-dmg' 'size:--disable-parallels' 'size:--disable-colo-proxy' 'size:--disable-debug-graph-lock' 'size:--disable-hexagon-idef-parser' 'size:--disable-libdw' 'size:--disable-pipewire' 'size:--disable-pixman' 'size:--disable-relocatable' 'size:--disable-rutabaga-gfx' 'size:--disable-vmdk' 'size:--disable-avx512bw' 'size:--disable-vpc' 'size:--disable-vhdx' 'size:--disable-hv-balloon' 'size:--disable-qpl' 'size:--disable-uadk' 'size:--disable-debug-remap' 'size:--disable-gio' 'size:--disable-libdaxctl' 'size:--disable-oss' 'speed:--enable-kvm' 'speed:--enable-vhost-net' 'size:--enable-linux-aio' 'size:--enable-linux-io-uring' 'functionality:--enable-virtfs' 'functionality:--enable-attr' 'functionality:--enable-cap-ng' 'functionality:--enable-seccomp' 'speed:--enable-avx2' 'speed:--enable-avx512bw' 'functionality:--disable-libpmem' 'speed:--enable-malloc-trim' 'arch:--target-list=x86_64-softmmu' 'functionality' 'security' 'speed' 'size:--extra-cflags=" -O2 -fno-semantic-interposition -falign-functions=32 -D_FORTIFY_SOURCE=2"' 'security:--extra-ldflags=" -z noexecstack -z relro -z now"' 'misc:--prefix=/opt/kata' 'arch:--libdir=/opt/kata/lib/kata-qemu' 'misc:--libexecdir=/opt/kata/libexec/kata-qemu' 'misc:--datadir=/opt/kata/share/kata-qemu')
```

zvonkok commented:
```
INFO: Found 0 patches
+ typeset -a qemu_options
+ typeset -A recognised_tags
+ prefix=/opt/kata
+ qemu_version=
+ recognised_tags=([arch]="architecture-specific" [functionality]="required functionality" [minimal]="specified to avoid building unnecessary elements" [misc]="miscellaneous" [security]="specified for security reasons" [size]="minimise binary size" [speed]="maximise startup speed")
+ main -s kata-qemu x86_64
+ action=
+ getopts dhms opt
+ case "$opt" in
+ static=true
+ getopts dhms opt
+ shift 1
+ '[' -z kata-qemu ']'
+ hypervisor=kata-qemu
+ local qemu_version_file=VERSION
+ '[' -f VERSION ']'
++ awk 'BEGIN {FS = "-"} {print $1}' VERSION
+ qemu_version=9.1.2
+ '[' -n 9.1.2 ']'
+ gt_eq 9.1.2 6.1.0
+ format='^[0-9]+(\.[0-9]+)*$'
+ [[ 9.1.2 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ [[ 6.1.0 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ echo -e '9.1.2\n6.1.0'
+ sort -V -r -C
++ gcc -dumpversion
++ cut -f1 -d.
+ local gcc_version_major=11
+ '[' -n 11 ']'
+ gt_eq 11 7.0.0
+ format='^[0-9]+(\.[0-9]+)*$'
+ [[ 11 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ [[ 7.0.0 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ echo -e '11\n7.0.0'
+ sort -V -r -C
++ gcc -dumpfullversion
++ cut -f2 -d.
+ local gcc_version_minor=4
+ '[' -n 4 ']'
+ local gcc_version=11.4
+ generate_qemu_options
++ uname -m
+ host=x86_64
+ '[' x86_64 '!=' x86_64 ']'
+ qemu_options+=(minimal:--without-default-features)
+ qemu_options+=(minimal:--without-default-devices)
+ qemu_options+=(size:--disable-brlapi)
+ qemu_options+=(minimal:--disable-docs)
+ qemu_options+=(size:--disable-curses)
+ qemu_options+=(size:--disable-gtk)
+ qemu_options+=(size:--disable-opengl)
+ qemu_options+=(size:--disable-sdl)
+ qemu_options+=(size:--disable-spice)
+ qemu_options+=(size:--disable-vte)
+ qemu_options+=(size:--disable-vnc)
+ qemu_options+=(size:--disable-vnc-jpeg)
+ gt_eq 9.1.2 7.0.50
+ format='^[0-9]+(\.[0-9]+)*$'
+ [[ 9.1.2 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ [[ 7.0.50 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ echo -e '9.1.2\n7.0.50'
+ sort -V -r -C
+ qemu_options+=(size:--disable-png)
+ qemu_options+=(size:--disable-vnc-sasl)
+ qemu_options+=(size:--disable-auth-pam)
+ qemu_options+=(size:--disable-glusterfs)
+ qemu_options+=(size:--disable-libiscsi)
+ qemu_options+=(size:--disable-libnfs)
+ qemu_options+=(size:--disable-libssh)
+ qemu_options+=(size:--disable-bzip2)
+ qemu_options+=(size:--disable-lzo)
+ qemu_options+=(size:--disable-snappy)
+ qemu_options+=(size:--disable-slirp)
+ qemu_options+=(size:--disable-libusb)
+ qemu_options+=(size:--disable-usb-redir)
+ case "$arch" in
+ qemu_options+=(size:--disable-tcg)
+ '[' true == true ']'
+ qemu_options+=(misc:--static)
+ case "$arch" in
+ qemu_options+=(size:--disable-debug-tcg)
+ qemu_options+=(size:--disable-tcg-interpreter)
+ qemu_options+=(size:--disable-qom-cast-debug)
+ qemu_options+=(size:--disable-libudev)
+ qemu_options+=(security:--disable-curl)
+ qemu_options+=(size:--disable-rdma)
+ qemu_options+=(size:--disable-tools)
+ gt_eq 9.1.2 8.0.0
+ format='^[0-9]+(\.[0-9]+)*$'
+ [[ 9.1.2 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ [[ 8.0.0 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ echo -e '9.1.2\n8.0.0'
+ sort -V -r -C
+ qemu_options+=(functionality:--enable-virtfs)
+ qemu_options+=(size:--disable-bsd-user)
+ qemu_options+=(size:--disable-linux-user)
+ qemu_options+=(size:--disable-sparse)
+ qemu_options+=(size:--disable-vde)
+ gt_eq 9.1.2 7.0.50
+ format='^[0-9]+(\.[0-9]+)*$'
+ [[ 9.1.2 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ [[ 7.0.50 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ echo -e '9.1.2\n7.0.50'
+ sort -V -r -C
+ qemu_options+=(size:--disable-nettle)
+ qemu_options+=(size:--disable-xen)
+ qemu_options+=(size:--disable-capstone)
+ qemu_options+=(size:--disable-virglrenderer)
+ qemu_options+=(size:--disable-replication)
+ qemu_options+=(size:--disable-smartcard)
+ qemu_options+=(size:--disable-guest-agent)
+ qemu_options+=(size:--disable-guest-agent-msi)
+ qemu_options+=(size:--disable-vvfat)
+ qemu_options+=(size:--disable-vdi)
+ qemu_options+=(size:--disable-qed)
+ qemu_options+=(size:--disable-qcow1)
+ qemu_options+=(size:--disable-bochs)
+ qemu_options+=(size:--disable-cloop)
+ qemu_options+=(size:--disable-dmg)
+ qemu_options+=(size:--disable-parallels)
+ qemu_options+=(size:--disable-colo-proxy)
+ qemu_options+=(size:--disable-debug-graph-lock)
+ qemu_options+=(size:--disable-hexagon-idef-parser)
+ qemu_options+=(size:--disable-libdw)
+ qemu_options+=(size:--disable-pipewire)
+ qemu_options+=(size:--disable-pixman)
+ qemu_options+=(size:--disable-relocatable)
+ qemu_options+=(size:--disable-rutabaga-gfx)
+ qemu_options+=(size:--disable-vmdk)
+ qemu_options+=(size:--disable-avx512bw)
+ qemu_options+=(size:--disable-vpc)
+ qemu_options+=(size:--disable-vhdx)
+ qemu_options+=(size:--disable-hv-balloon)
+ gt_eq 9.1.2 9.1.0
+ format='^[0-9]+(\.[0-9]+)*$'
+ [[ 9.1.2 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ [[ 9.1.0 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ echo -e '9.1.2\n9.1.0'
+ sort -V -r -C
+ qemu_options+=(size:--disable-qpl)
+ qemu_options+=(size:--disable-uadk)
+ qemu_options+=(size:--disable-debug-remap)
+ qemu_options+=(size:--disable-gio)
+ qemu_options+=(size:--disable-libdaxctl)
+ qemu_options+=(size:--disable-oss)
+ '[' true == true ']'
+ '[' x86_64 == aarch64 ']'
+ qemu_options+=(speed:--enable-kvm)
+ qemu_options+=(speed:--enable-vhost-net)
+ qemu_options+=(size:--enable-linux-aio)
+ qemu_options+=(size:--enable-linux-io-uring)
+ '[' -z true ']'
+ qemu_options+=(functionality:--enable-virtfs)
+ qemu_options+=(functionality:--enable-attr)
+ qemu_options+=(functionality:--enable-cap-ng)
+ qemu_options+=(functionality:--enable-seccomp)
+ '[' x86_64 == x86_64 ']'
+ qemu_options+=(speed:--enable-avx2)
+ qemu_options+=(speed:--enable-avx512bw)
+ qemu_options+=(functionality:--disable-libpmem)
+ qemu_options+=(speed:--enable-malloc-trim)
+ '[' x86_64 = ppc64le ']'
+ qemu_options+=(arch:"--target-list=${arch}-softmmu")
+ '[' -z true ']'
+ _qemu_cflags=
+ _qemu_cflags+=' -O2'
+ gt_eq 11.4 5.3.0
+ format='^[0-9]+(\.[0-9]+)*$'
+ [[ 11.4 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ [[ 5.3.0 =~ ^[0-9]+(\.[0-9]+)*$ ]]
+ echo -e '11.4\n5.3.0'
+ sort -V -r -C
+ _qemu_cflags+=' -fno-semantic-interposition'
+ _qemu_cflags+=' -falign-functions=32'
+ _qemu_cflags+=' -D_FORTIFY_SOURCE=2'
+ qemu_options+=(functionality,security,speed,size:"--extra-cflags=\"${_qemu_cflags}\"")
+ unset _qemu_cflags
+ _qemu_ldflags=
+ _qemu_ldflags+=' -z noexecstack'
+ _qemu_ldflags+=' -z relro'
+ _qemu_ldflags+=' -z now'
+ qemu_options+=(security:"--extra-ldflags=\"${_qemu_ldflags}\"")
+ unset _qemu_ldflags
+ qemu_options+=(misc:--prefix=${prefix})
+ qemu_options+=(arch:--libdir=${prefix}/lib/${hypervisor})
+ qemu_options+=(misc:--libexecdir=${prefix}/libexec/${hypervisor})
+ qemu_options+=(misc:--datadir=${prefix}/share/${hypervisor})
```

zvonkok commented:
|  command  |   time [s]   | branch   |  objects | 
| --------------- | --------- | ------------ | ---------| 
| ❯ time USE_CACHE=no  make qemu-tarball  | make: 16   | main | 2085 |
| ❯ time USE_CACHE=no  make qemu-tarball  | make: 12  | PR10708 |

zvonkok commented on tools/packaging/static-build/qemu/build-qemu.sh:
```diff
@@ -37,6 +37,12 @@ if [[ "$(uname -m)" != "${ARCH}" ]] && [[ "${ARCH}" == "s390x" ]]; then
        PREFIX="${PREFIX}" "${kata_packaging_scripts}"/configure-hypervisor.sh -s "${HYPERVISOR_NAME}" "${ARCH}" | xargs ./configure  --with-pkgversion="${PKGVERSION}"
 fi
+
+[[ ${ARCH} == "x86_64"   ]] && echo 'CONFIG_Q35=y'             >> configs/devices/i386-softmmu/default.mak
```
Holy moly :)

zvonkok commented:
Local build comparison on aarch64 (144 cores), using the same GCC 11.4 builder environment for both.

| | main (v11.0.0) | this PR (v11.0.1) | delta |
|---|---|---|-:|
| `qemu-system-aarch64` | 108 MB (unstripped) | 29 MB (stripped) | **−73%** |
| tarball (`.tar.zst`) | 50 MB | 20 MB | **−60%** |
| build time (from source) | _(ORAS cache hit)_ | 4m 36s (144 cores) | — |

Main was pulled from the ORAS artifact cache rather than compiled locally; the branch was built from source.

The binary size delta comes from a combination of changes introduced in this PR:

- `--without-default-features` / `--without-default-devices` — allnoconfig baseline, only Kata-required devices compiled in
- single targeted binary (`qemu-system-aarch64` only, no tools or tests)
- `--enable-strip` — debug symbols removed at install time (main ships an unstripped binary)
- `--enable-lto` — link-time optimisation for dead-code elimination across translation units
- `--disable-vmdk` and other storage formats not needed for Kata workloads

The firmware blacklist (architecture-appropriate pruning) contributes to the tarball reduction.

fidencio commented on tools/packaging/scripts/configure-hypervisor.sh:
```diff
@@ -389,15 +396,16 @@ generate_qemu_options() {
 	qemu_options+=(size:--disable-relocatable)
 	qemu_options+=(size:--disable-rutabaga-gfx)
-	if ! gt_eq "${qemu_version}" "10.1.0" ; then
+	qemu_options+=(size:--disable-vmdk)
```
Disabling VMDK here is part of the idea, but we never re-enable it (and it's needed for erofs).

Apokleos replied:
agree with @fidencio we need vmdk enabled in qemu compiling for erofs

zvonkok replied:
Fixed: `--enable-vmdk` is restored (de157e727) with the EROFS fsmerge comment, and the stray `--disable-vmdk` is dropped (bf5f009db) so the requirement is expressed once, unambiguously.

fidencio commented on tools/packaging/scripts/configure-hypervisor.sh:
```diff
@@ -451,32 +469,22 @@ generate_qemu_options() {
 	# Enable libc malloc_trim() for memory optimization.
 	qemu_options+=(speed:--enable-malloc-trim)
 
-	# Required by multi-layer EROFS rootfs with fsmerge (VMDK flat-extent descriptor)
-	qemu_options+=(functionality:--enable-vmdk)
-
```
Here is where we used to re-enable it, but it got fully dropped.

zvonkok replied:
Restored in de157e727, with the comment explaining the multi-layer EROFS fsmerge dependency (VMDK flat-extent descriptor).

fidencio reviewed (commented):
Let me comment the issues I found along the way, starting with VMDK support being needed for EROFS support.

fidencio commented on tools/packaging/scripts/configure-hypervisor.sh:
```diff
@@ -423,12 +431,25 @@ generate_qemu_options() {
+	# removes code paths that survive individual compilation but become
+	# unreachable across the final link.  With our minimal device set this
+	# typically shrinks the binary by 15-25%.  Build time increases ~2-3x.
+	qemu_options+=(size:--enable-lto)
```
I'm not sure if it can be done for all the architectures, as:
```
FAILED: qemu-system-ppc64
string_fortified.h:29:10: error: '__builtin_memcpy' writing 16 bytes into a region of size 0 overflows the destination [-Werror=stringop-overflow=]
lto1: all warnings being treated as errors
```

zvonkok replied:
Agreed - LTO is now guarded to `arch != ppc64le` (7f173e397). ppc64le builds without LTO; the other architectures keep the 15-25% size win.

fidencio commented on tools/packaging/scripts/configure-hypervisor.sh:
```diff
@@ -437,9 +458,6 @@ generate_qemu_options() {
 	# Support Ceph RADOS Block Device (RBD)
 	[[ -z "${static}" ]] && qemu_options+=(functionality:--enable-rbd)
 
-	# Support NUMA topology
-	qemu_options+=(functionality:--enable-numa)
-
```
NUMA got dropped entirely and was not re-enabled.

zvonkok replied:
Fixed: `--enable-numa` restored in de157e727.

Apokleos commented:
> vmdk support                                 : NO

We need VMDK for erofs

fidencio commented on tools/packaging/scripts/configure-hypervisor.sh:
```diff
@@ -213,14 +212,18 @@ generate_qemu_options() {
 
+	# We're starting with the minimal set of options required to build
+	qemu_options+=(minimal:--without-default-features)
+	#qemu_options+=(minimal:--without-default-devices)
```
So, for now we're not getting into the business of disabling devices, only features.

We should make it clear in the commit message.

zvonkok replied:
Update: device disablement is now active in this PR - 37b3bfe34 turns on `--without-default-devices` together with the completed per-architecture allowlist, and the commit message spells out the scope (features AND devices) and where the allowlist lives.

Apokleos replied:
Shall we directly fix this in the commit but not in `37b3bfe`

fidencio commented on tools/packaging/static-build/qemu/build-qemu.sh:
```diff
@@ -27,20 +18,94 @@ kata_static_build_scripts="${kata_static_build_dir}/scripts"
+trap 'rm -rf "${workdir}"' EXIT
+git clone --depth=1 "${QEMU_REPO}" "${workdir}/qemu"
+pushd "${workdir}/qemu"
 git fetch --depth=1 origin "${QEMU_VERSION_NUM}"
 git checkout FETCH_HEAD
 scripts/git-submodule.sh update meson capstone
-"${kata_packaging_scripts}/patch_qemu.sh" "${QEMU_VERSION_NUM}" "${kata_packaging_dir}/qemu/patches"
+"${kata_packaging_scripts}"/patch_qemu.sh "${QEMU_VERSION_NUM}" "${kata_packaging_dir}/qemu/patches"
+
+# With --without-default-devices every machine type and every device with
+# "default y" is suppressed (allnoconfig semantics).  We must explicitly
+# list each CONFIG_* we need before ./configure so meson picks them up.
+#
+# Common virtio devices used by Kata on every architecture:
+#   VIRTIO_BLK  – rootfs / container block device
+#   VIRTIO_NET  – container networking (accelerated by vhost-kernel)
+#   VIRTIO_SERIAL / VIRTIO_CONSOLE – agent tty / console
+#   VIRTIO_RNG  – guest entropy source
+#   VIRTIO_BALLOON – memory pressure management
+#   VIRTIO_MEM  – memory hot-plug
+#   VHOST_USER_FS – virtiofsd shared filesystem (vhost-user)
+#   VIRTIO_9P   – fallback 9P shared filesystem
+#   VHOST_VSOCK – VM↔host socket for the Kata agent (kernel vhost)
+#   VFIO_PCI    – GPU / NIC passthrough via VFIO (also selects VFIO)
+#   IOMMUFD     – modern VFIO backend (depends on VFIO)
+
+_COMMON_DEVS='
+CONFIG_VIRTIO_PCI=y
+CONFIG_VIRTIO_BLK=y
+CONFIG_VIRTIO_NET=y
+CONFIG_VIRTIO_SERIAL=y
+CONFIG_VIRTIO_RNG=y
+CONFIG_VIRTIO_BALLOON=y
+CONFIG_VIRTIO_MEM=y
+CONFIG_VHOST_USER_FS=y
+CONFIG_VIRTIO_9P=y
+CONFIG_VHOST_VSOCK=y
+CONFIG_VFIO_PCI=y
+CONFIG_IOMMUFD=y
+'
```
As we're not disabling the default devices yet we should either:
* Remove this and land it in a follow-up PR where we actually do the device disablement by default
* Do the device disablement as part of this PR

I'd recommend the first.

zvonkok replied:
We went with the second option: 37b3bfe34 activates `--without-default-devices`. The reason the flag had been parked - an incomplete allowlist - is fixed and now *guarded*:

- The allowlist covers every device both runtimes can emit, including the previously missing ones: virtio-scsi/scsi-hd, nvdimm, pcie-root-port, x3130-upstream/xio3130-downstream, pci-bridge, pcie-pci-bridge, pxb-pcie (x86), vhost-user-blk/scsi, pvpanic.
- A post-build `verify_devices` gate runs `qemu-system-* -device help` and fails the build if any required device is missing, so an allowlist gap fails at build time instead of at VM start.
- Validated with a native aarch64 static build: all 22 required devices present (incl. arm-smmuv3, pxb-pcie, xio3130-downstream, nvdimm), stripped binary 23.5 MB, kata-static-qemu.tar.gz 6.4 MB.

fidencio commented on tools/packaging/static-build/qemu/build-qemu.sh:
```diff
@@ -27,20 +18,94 @@ kata_static_build_scripts="${kata_static_build_dir}/scripts"
+		>> configs/devices/ppc64-softmmu/default.mak
+	unset _PPC64_DEVS
+fi
+unset _COMMON_DEVS _S390_DEVS
```
Same comment applies about the arch specific devices.

zvonkok replied:
The arch-specific entries stay and are now actually exercised: the flag is active (37b3bfe34) and the post-build `verify_devices` gate proves each architecture's device set is complete at build time.

fidencio reviewed (commented):
Some comments about `--without-default-devices` not being wired here.

manuelh-dev commented on tools/packaging/scripts/configure-hypervisor.sh:
```diff
@@ -188,11 +188,10 @@ show_array() {
 		if [[ "${action}" = "dump" ]]; then
```
a general thought for this script - it seems generally misplaced - this is qemu specific, could be called configure-qemu.sh and moved to tools/packaging/static-build/qemu

manuelh-dev replied:
also, at first glance, tools/packaging/static-build/scripts/qemu-build-post.sh seems misplaced. Should it live under the qemu folder as build-qemu-post.sh?

I don't want to stir things up here for this very PR - just bringing this to attention as you're generally working on QEMU.

manuelh-dev replied:
@zvonkok for your visibility as part of the qemu refactor possibly ... or not ...

zvonkok commented on tools/packaging/static-build/qemu/build-qemu.sh:
```diff
@@ -18,19 +18,93 @@ kata_static_build_scripts="${kata_static_build_dir}/scripts"
+
+rm -rf qemu
 git clone --depth=1 "${QEMU_REPO}" qemu
 pushd qemu
```
Fixed in cc93a64af: the clone now goes into a `mktemp -d` directory with a `trap` cleanup, so nothing in the bind-mounted host working directory can be touched.

zvonkok commented on tools/packaging/static-build/qemu/build-qemu.sh:
```diff
@@ -18,19 +18,93 @@ kata_static_build_scripts="${kata_static_build_dir}/scripts"
+elif [[ "${ARCH}" == "ppc64le" ]]; then
+	# VIRTIO_MEM depends on VIRTIO_MEM_SUPPORTED, which is only selected on
+	# arm/i386/s390x — ppc64 PSeries does not support virtio-mem.
+	_PPC64_DEVS=$(printf '%s' "${_COMMON_DEVS}" | grep -v 'VIRTIO_MEM\b')
```
Fixed in df116fe4b: the filter now matches the exact config prefix (`grep -v 'CONFIG_VIRTIO_MEM='`), no PCRE-only escapes.

zvonkok commented on tools/packaging/scripts/configure-hypervisor.sh:
```diff
@@ -389,15 +400,15 @@ generate_qemu_options() {
-	if ! gt_eq "${qemu_version}" "10.1.0" ; then
+	if ! gt_eq "${qemu_version}" "10.1.0" && [[ "${arch}" == x86_64 ]]; then
 		qemu_options+=(size:--disable-avx512bw)
 	fi
```
Fixed in bb0d77c2d: the `--disable-avx512bw` is dropped; the speed-tagged `--enable-avx512bw` (which won by ordering anyway) is now the only mention.

zvonkok commented on tools/packaging/scripts/configure-hypervisor.sh:
```diff
@@ -571,7 +589,8 @@ main() {
 			;;
 		*)
-			die "Invalid option: -${OPTARG:-${opt}}"
+			usage
+			exit 1
 			;;
```
Fixed in 31cb99d2a: restored `die "Invalid option: -..."` matching the behavior on main.

zvonkok commented on tools/packaging/scripts/lib.sh:
```diff
@@ -187,7 +187,13 @@ calc_qemu_files_sha256sum() {
 
 get_qemu_image_name() {
 	qemu_script_dir="${repo_root_dir}/tools/packaging/static-build/qemu"
-	echo "${BUILDER_REGISTRY}:qemu-$(get_last_modification "${qemu_script_dir}")-$(uname -m)"
+	qemu_configure_dir="${repo_root_dir}/tools/packaging/scripts"
+
+	qemu_hash=$(merge_two_hashes \
+		"$(get_last_modification "${qemu_script_dir}")" \
+		"$(get_last_modification "${qemu_configure_dir}")")
+
```
Fixed in 7c3a3d2d1: the tag now hashes only `configure-hypervisor.sh` instead of the whole scripts directory.

zvonkok commented:
All review comments are addressed as of 7c3a3d2d1:

- **VMDK** (@fidencio, @Apokleos): `--enable-vmdk` restored (de157e727) and the stray `--disable-vmdk` dropped (bf5f009db): `vmdk support: YES` in the configure summary again.
- **NUMA**: `--enable-numa` restored (de157e727).
- **LTO**: guarded to `arch != ppc64le` (7f173e397): ppc64 fails under LTO with stringop-overflow errors.
- **Device disablement**: now *active* in this PR (37b3bfe34). `--without-default-devices` is on, the per-arch allowlist covers every device both runtimes can emit (the previously missing virtio-scsi/scsi-hd, nvdimm, pcie-root-port, x3130/xio3130, pci-bridge, pcie-pci-bridge, pxb-pcie, vhost-user-blk/scsi, pvpanic included), and a post-build `verify_devices` gate greps `qemu-system-* -device help` and **fails the build** if anything kata needs is missing: allowlist gaps surface at build time, not VM start.

Updated numbers with devices disabled, native aarch64 static build (same 144-core builder):

| | main (v11.0.0) | this PR (v11.0.1) | delta |
|---|---|---|-:|
| `qemu-system-aarch64` | 108 MB (unstripped) | 23.5 MB (stripped) | **−78%** |
| tarball (`.tar.gz`) | 50 MB | 6.4 MB | **−87%** |

`verify_devices: all 22 required devices present in ./build/qemu-system-aarch64`

manuelh-dev commented on tools/packaging/static-build/qemu/build-qemu.sh:
```diff
@@ -27,20 +18,187 @@ kata_static_build_scripts="${kata_static_build_dir}/scripts"
+	echo "verify_devices: all $# required devices present in ${qemu_binary}"
+}
+
+_pci_devs=(
```
I would define _pci_devs and _pcie_topology closer to _COMMON_DEVS and _PCIE_DEVS as well as the input to verify_devices for each arch closer to where we generate default.mak. Makes it a bit easier to maintain, read, adjust understand, ...

manuelh-dev commented on tools/packaging/static-build/qemu.blacklist:
```diff
@@ -33,3 +33,97 @@ qemu_black_list=(
+#   aarch64 -> edk2-aarch64-*.fd (AAVMF)
+#   s390x   -> s390-ccw.img (CCW BIOS)
+#   ppc64le -> slof.bin / vof.bin (Open Firmware)
+ARCH=${ARCH:-$(uname -m)}
```
hmm - I was trying to understand where there could be issues with ARCH variable propagation or cross-compilation. 

I think: 
- we should remove the sourcing of this file in tools/packaging/static-build/qemu/build-base-qemu.sh - it is not needed there is it?
- reconsider use of the ARCH variable in tools/packaging/static-build/scripts/qemu-build-post.sh ... it just uses ARCH. Possibly ARCH is defined because we source qemu.blacklist? 
  - maybe qemu-build-post.sh should set or assure ARCH=${ARCH:-$(uname -m)} before sourcing the blacklist

manuelh-dev commented on tools/packaging/static-build/qemu.blacklist:
```diff
@@ -33,3 +33,97 @@ qemu_black_list=(
+#   aarch64 -> edk2-aarch64-*.fd (AAVMF)
+#   s390x   -> s390-ccw.img (CCW BIOS)
+#   ppc64le -> slof.bin / vof.bin (Open Firmware)
+ARCH=${ARCH:-$(uname -m)}
```
Also, honestly, I am not sure if all our ARCH definitions using `uname -m` work under cross-compilation. Maybe since these are fallbacks like in this file, the fallback assignment is never triggered. But I would recommend to revisit setting these ARCH variables with cross-compilation in mind

manuelh-dev replied:
addressed

fidencio commented:
@zvonkok, I've added a fixup that solves build issues on x86_64 (due to missing CXL core deps) and on s390x (due to missing VFIO deps). Let's see how CI goes.

fidencio commented:
rebasing due to the trustee & guest components bump.

fidencio commented:
We need rebase, doing it now.

zvonkok commented:
This work is related to this effort: https://github.com/kata-containers/kata-containers/issues/13424

Apokleos commented on tools/packaging/static-build/qemu/build-qemu.sh:
```diff
@@ -26,24 +26,84 @@ HYPERVISOR_NAME="${HYPERVISOR_NAME:-}"
 PREFIX="${PREFIX:-}"
 PKGVERSION=${PKGVERSION:-}
 
+rm -rf qemu
```
I notice this command was removed in the commit e2b9125289ca90660e9e830d830346e2696927b7 but now it is added again.

fidencio replied:
@Apokleos, I've taken your comment as an opportunity to re-organise the PRs and fold things together.
Re-pushed.

Apokleos commented on tools/packaging/static-build/qemu/build-qemu.sh:
```diff
@@ -26,11 +26,10 @@ HYPERVISOR_NAME="${HYPERVISOR_NAME:-}"
 
-# This script always runs inside a dedicated build container (see
-# static-build/qemu/Dockerfile), so rm -rf of a relative path is safe.
-rm -rf qemu
```
I remember these lines were added in the prior commits but why remove them here ?

fidencio replied:
@Apokleos, I've taken your comment as an opportunity to re-organise the PRs and fold things together.
Re-pushed.

Apokleos reviewed (approved):
Thx @zvonkok LGTM, but some comments left to figure out that we'd better fix the comments in place.

fidencio commented:
Converted to draft to address @Apokleos' review.

zvonkok committed:
qemu: drop --disable-avx512bw conflicting with the later enable

For x86_64 on QEMU < 10.1.0 the option list carries both
--disable-avx512bw and --enable-avx512bw under identical conditions (the
enable wins by ordering, so AVX512BW is built in).  On the remaining
architectures the option is meaningless: avx512bw selects an x86 host
code path.

Remove the disable so the speed-tagged enable is the single, unambiguous
expression of the requirement.

Signed-off-by: Zvonko Kaiser <zkaiser@nvidia.com>

zvonkok committed:
shellcheck: fix configure-hypervisor.sh

Write the two constructs that needed a `shellcheck disable` in a form
shellcheck accepts (the line-continuation suffix and the comma-tagged
--extra-cflags entry), quote the remaining expansions and normalise the
formatting of the conditionals and case arms.

No functional change: the option list generated for every architecture
and every supported QEMU version is byte-identical before and after.

Signed-off-by: Zvonko Kaiser <zkaiser@nvidia.com>

zvonkok committed:
qemu: Update get_qemu_image_name

The QEMU builder image tag only tracked the modification time of
static-build/qemu, so a change to configure-hypervisor.sh - the script
that shapes the whole QEMU build - reused a stale builder image.

Fold the modification time of configure-hypervisor.sh into the tag as
well.  Only that script is hashed, not the entire packaging scripts
directory, so unrelated packaging changes do not invalidate the image.

Signed-off-by: Zvonko Kaiser <zkaiser@nvidia.com>

zvonkok committed:
qemu: Clone into a temporary directory

The script runs with the container working directory bind-mounted from
the host, so `rm -rf qemu` deletes a host directory whenever the caller
happens to have a qemu/ in the tree it builds from.

Clone into a mktemp -d directory instead and remove it from a trap, so
the build only ever writes inside its own scratch space and cleans up on
every exit path.

Signed-off-by: Zvonko Kaiser <zkaiser@nvidia.com>

zvonkok committed:
qemu: Add machines and devices for each architecture

--without-default-devices uses allnoconfig semantics: every Kconfig entry
with "default y" is suppressed, which drops every machine type and every
virtio device from the binary.  Write the set Kata actually needs into
each architecture's default.mak before ./configure runs, so meson picks
it up.

The list is split into transport-independent device models (_COMMON_DEVS,
all architectures) and the PCI transport plus PCIe slot topology
(_PCIE_DEVS, x86_64/aarch64).  ppc64le takes the conventional PCI subset,
since pSeries PHBs have no PCIe ports, and drops VIRTIO_MEM, which
pSeries does not support; s390x is pure CCW.  Every device the Go runtime
and runtime-rs can emit is covered, including virtio-scsi/scsi-hd, the
vhost-user backends, nvdimm, the root/switch/bridge port models, pxb-pcie
and pvpanic, plus the TDX, SEV and SGX objects, which QEMU only implies
from CONFIG_PC.

Build just the system emulator rather than the full tool set, and gate
the result: verify_devices and verify_objects run the freshly built
binary with -device help and -object help and fail the build if anything
on the list is missing.  An allowlist gap therefore surfaces here instead
of at VM start.

This is inert until --without-default-devices is passed, which the next
commit does.

Signed-off-by: Zvonko Kaiser <zkaiser@nvidia.com>

zvonkok committed:
qemu: Disable features/devices per default

We carry a very long list of --disable-* calls that the configure script
can express in two options: --without-default-features and
--without-default-devices turn everything off, and we ask back only for
what Kata uses.  The devices we ask back for are the per-architecture
allowlist added in the previous commit, which the build verifies against
the resulting binary.

Both flags imply auto_features=disabled, which also silently switches off
the vhost-kernel and vhost-user backends: vhost-net acceleration and
virtiofsd would keep working from the runtime's point of view right up to
VM start.  Enable both explicitly.

Signed-off-by: Zvonko Kaiser <zkaiser@nvidia.com>

zvonkok committed:
qemu: reduce attack surface and binary size

Strip the installed binaries and build them with LTO, which pays off
particularly well now that the device set is small: cross-unit dead-code
elimination removes what only becomes unreachable at link time.  ppc64le
opts out, as LTO promotes a __builtin_memcpy bound check in
qemu-system-ppc64 to a -Werror=stringop-overflow failure.

Also record why aarch64 still builds TCG while every other architecture
disables it, so the gap is visible rather than looking accidental.

Signed-off-by: Zvonko Kaiser <zkaiser@nvidia.com>

zvonkok committed:
qemu: prune non-applicable firmware from tarball

QEMU installs firmware blobs for every platform it knows about, not just
the one it was built for.  Keep only what the target architecture can
actually boot - OVMF on x86_64, AAVMF on aarch64, s390-ccw.img on s390x,
SLOF/VOF on ppc64le - and blacklist the rest at packaging time, so the
tarball carries neither the extra size nor executable blobs that have no
function on the host.

Signed-off-by: Zvonko Kaiser <zkaiser@nvidia.com>

manuelh-dev reviewed (approved):
Thanks, LGTM, pending last ongoing CI runs.
